standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/PropertyManager.java

```diff
@@ -0,0 +1,629 @@
+      UUID digest = store.fetchDigest(mapKey);
+      // not stored nor cached, nothing to do
+      if (digest == null) {
+        return false;
+      }
+      map = new PropertyMap(store, schemaOf(splitKey(mapName + ".*")), PropertyMap.DROPPED);
+      synchronized (dirtyMaps) {
+        dirtyMaps.put(mapName, map);
+      }
+    }
+    // if this is the first update to the map
+    if (map != dirtyMap) {
+      dirtyMaps.put(mapName, map);
+    }
+    return false;
+  }
+
+  /**
+   * Sets a property value.
+   * @param keys the key fragments
+   * @param value the new value or null if mapping should be removed
+   */
+  protected void setProperty(String[] keys, Object value) {
+    // find schema from key (length)
+    PropertySchema schema = schemaOf(keys);
+    String mapKey = mapKey(keys);
+    PropertyMap dirtyMap;
+    final Map<String, PropertyMap> dirtyMaps = this.dirtyMaps;
+    synchronized (dirtyMaps) {
+      dirtyMap = dirtyMaps.get(mapKey);
+    }
+    PropertyMap map;
+    if (dirtyMap != null) {
+      map = dirtyMap;
+    } else {
+      // is is stored ?
+      map = store.fetchProperties(mapKey, s->schema);
+      if (map == null) {
+        // remove a value from a non persisted map, noop
+        if (value == null) {
+          return;
+        }
+        map = new PropertyMap(store, schema);
+      }
+    }
+    // map is not null
+    String propertyName = propertyName(keys);
+    if (value != null) {
+      map.putProperty(propertyName, value);
+    } else {
+      map.removeProperty(propertyName);
+    }
+    // if this is the first update to the map
+    if (map != dirtyMap) {
+      dirtyMaps.put(mapKey, map);
+    }
+  }
+
+  /**
+   * Selects a set of properties.
+   * @param namePrefix the map name prefix
+   * @param predicateStr the condition selecting maps
+   * @param projectStr the projection property names or script
+   * @return the map of property maps keyed by their name
+   */
+  public Map<String, PropertyMap> selectProperties(String namePrefix, String predicateStr, String... projectStr) {
+    return selectProperties(namePrefix, predicateStr,
+        projectStr == null
+            ? Collections.emptyList()
+            : Arrays.asList(projectStr));
+  }
+
+  /**
+   * Selects a set of properties.
+   * @param namePrefix the map name prefix
+   * @param selector the selector/transformer function
+   * @return the map of property maps keyed by their name
+   */
+  public Map<String, PropertyMap> selectProperties(String namePrefix, Function<PropertyMap, PropertyMap> selector) {
+    final String mapKey = mapKey(namePrefix);
+    final Map<String, PropertyMap> selected = store.selectProperties(mapKey,null, k->schemaOf(splitKey(k)) );
+    final Map<String, PropertyMap> maps = new TreeMap<>();
+    final Function<PropertyMap, PropertyMap> transform = selector == null? Function.identity() : selector;
+    selected.forEach((k, p) -> {
+      final PropertyMap dirtyMap = dirtyMaps.get(k);
```

Comment:
nit: If I understand correctly, the `dirtyMaps` will always be empty for `selectProperties` method

Reply:
They are not in the case of a script where put properties can be followed by select in the same tx. Adding a test (and fixing a bug for new maps).

Reply:
For every new select request, we always create a new PropertyManager, so I guess the `dirtyMaps` is always empty in this way.